UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Curtailment/BMRS-API-JA-PerGenerationUnitCurtailment-Bulk-2.py

```diff
@@ -0,0 +1,154 @@
+#KEY: iwx6raw9m7nqq0f
+
+#TEST from page 193: https://www.elexon.co.uk/documents/training-guidance/bsc-guidance-notes/bmrs-api-and-data-push-user-guide-2/
+
+###Libraries###
+import httplib2
+from pprint import pformat
+
+import re
+import pandas as pd
+import sys
+from datetime import datetime, timedelta
+
+
+####Functions###
+def str0_2(value):
+    #Converts the day or month int to a string of length 2. Thus, 12 -> "12", and 1 -> "01", the leading 0 is important.
+    #This function thus performs a similar role as str(), but also can add the 0 at the start, and is applied to length 2 instances.
+    #Must use for day or month, but use for period is optional, as length 1 and 2 is accepted by the API format the period, but not the month or day. 
+    value = str(value)
+    if len(value) == 1:
+        value = "0"+value
+    return value
+
+
+def next_day(Year, Month, Day):
+    #Convert to timestamp.
+    day = datetime(int(Year), int(Month), int(Day), 0, 0, 0, 0)
+    
+    #Get a week ago (returns str for year, month, day: 
+    nextDay = day + timedelta(days=1)
+    print(nextDay)
+    return str(nextDay.year), str0_2(nextDay.month), str0_2(nextDay.day)
+
+
+def post_elexon(url):
+    http_obj = httplib2.Http()
+    resp, content = http_obj.request(
+        uri=url,
+        method='GET',
+        headers={'Content-Type': 'application/xml; charset=UTF-8'},
+    )
+    #print('===Response===')
+    #print(pformat(resp))
+    
+    #print('===Content===')
+    #print(pformat(content))
+    #print(type(pformat(content)))
+
+    #print('===Finished===')
+    
+    return pformat(content)
+
+
+def query_curtailment(YYYY, MM, DD, PP):
+    #NOTE: IF YOU WANT PRICES IN THE FUTURE, then I reccomend duplicating this function, but going by bidPrice (or something else like that once you've checked what is needed) instead of bidVolume.
+    #Keep in mind, however, that there's a but and sell side to keep track of (think the buy side might use something like offerPrice and offerVolume, rather than bidPrice and bidVolume), so you'd need to consider this change too.
+    #Currently the bidding is not used here, but rather, as this is for curtailment, sell side volumes are considered.
+    #This process is also specifically used for renewables. 
+    queryoutput = post_elexon(url='https://api.bmreports.com/BMRS/DETSYSPRICES/v1?APIKey=iwx6raw9m7nqq0f&SettlementDate='+YYYY+'-'+MM+'-'+DD+'&SettlementPeriod='+PP+'&ServiceType=xml',) #All
```

Comment:
Please remove the API key value from lines like this. Don't remove the lines as a whole.

Reply:
Done.